kpops/utils/gen_schema.py

```diff
@@ -76,32 +76,59 @@ def is_valid_component(
         return True
 
 
-def gen_pipeline_schema(
-    components_module: str | None = None,
-) -> None:
+def _add_components(components_module: str, components: tuple = ()) -> tuple:
```

Comment:
I am now overwhelmed by mypy errors of all kinds :X

Reply:
I think the type checker doesn't really like the empty tuple generation as `()`, if I change it to `tuple()` it passes for Pylance

Reply:
btw I am getting
```
realpath: --canonicalize-existing: No such file or directory
```

when I run `pre-commit run -a mypy`

Reply:
Hmm, you are using `zsh`, right? Could that be the reason?

Reply:
`mypy` doesn't like the tuple being empty at all it seems.

Reply:
It is also concerned about the `components` being none despite the logical check.

Reply:
that's weird, it sounds like it doesn't perform type narrowing correctly or it has a very strict idea of how tuples should be typed/used. could you check if there are any related tickets on the mypy issue tracker?

Reply:
I would use the `assert` statement, but I am still left with the following:

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

kpops/utils/gen_schema.py:96: error: Incompatible types in assignment
(expression has type "Tuple[<nothing>, ...]", variable has type
"Optional[Tuple[Type[PipelineComponent]]]")  [assignment]
            components = tuple()
                         ^~~~~~~
kpops/utils/gen_schema.py:110: error: Incompatible types in assignment
(expression has type "Tuple[Type[PipelineComponent], ...]", variable has type
"Optional[Tuple[Type[PipelineComponent]]]")  [assignment]
        components += tuple(custom_components)
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
kpops/utils/gen_schema.py:137: error: Argument 2 to "_add_components" has
incompatible type "Tuple[Type[PipelineComponent], ...]"; expected
"Optional[Tuple[Type[PipelineComponent]]]"  [arg-type]
            components = _add_components(components_module, components)
                                                            ^~~~~~~~~~
Found 3 errors in 1 file (checked 109 source files)

```

Why would narrowing the type be bad? Besides, I am asserting that the tuple is not None and then mypy still thinks it is of type Optional??

Reply:
I would use the `assert` statement, but I am still left with the following:

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

kpops/utils/gen_schema.py:96: error: Incompatible types in assignment
(expression has type "Tuple[<nothing>, ...]", variable has type
"Optional[Tuple[Type[PipelineComponent]]]")  [assignment]
            components = tuple()
                         ^~~~~~~
kpops/utils/gen_schema.py:110: error: Incompatible types in assignment
(expression has type "Tuple[Type[PipelineComponent], ...]", variable has type
"Optional[Tuple[Type[PipelineComponent]]]")  [assignment]
        components += tuple(custom_components)
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
kpops/utils/gen_schema.py:137: error: Argument 2 to "_add_components" has
incompatible type "Tuple[Type[PipelineComponent], ...]"; expected
"Optional[Tuple[Type[PipelineComponent]]]"  [arg-type]
            components = _add_components(components_module, components)
                                                            ^~~~~~~~~~
Found 3 errors in 1 file (checked 109 source files)

```

I am asserting that the tuple is not None and then mypy still thinks it is of type Optional??
Also, why is passing a tuple as an arg, instead of an optional tuple flagged?

Reply:
I could silence all errors, but I don;t understand why `mypy` handles tuples like that

Reply:
yeah, that's very unfortunate